### PR TITLE
Start reading from new offer and conditions tables

### PR DESCRIPTION
**Pull request by despo**
## Context
Read offer data from Offer and OfferConditions in the Manage UI.

## Link to Trello card

https://trello.com/c/x9taO1pU/3731-%F0%9F%8F%88-start-reading-from-new-offer-and-conditions-tables

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/controllers/vendor_api/decisions_controller.rb [hunk 1]
```diff
@@ -67,7 +67,7 @@ def application_choice
     end
 
     def render_application
-      render json: { data: SingleApplicationPresenter.new(application_choice).as_json }
+      render json: { data: SingleApplicationPresenter.new(application_choice.reload).as_json }
```
Comment (benswannack) [inline app/controllers/vendor_api/decisions_controller.rb:70]: Why is this needed?

**Review by benswannack — COMMENTED** [review]
Still going through, but some initial comments here

File: app/controllers/vendor_api/decisions_controller.rb (hunk 1 above)
Comment (despo) [reply app/controllers/vendor_api/decisions_controller.rb:70]: Because the application_choice object is not updated as part of Make or ChangeOffer service to include `offer` (as the association is not setup and the offer object is created separately), which results in the API returning out of date information. We can remove when the association is implemented.

File: app/models/application_choice.rb [hunk 2]
```diff
@@ -78,6 +78,10 @@ def days_until_decline_by_default
     end
   end
 
+  def offer
```
Comment (benswannack) [inline app/models/application_choice.rb:81]: What was the issue with defining the association for `offer`?

File: app/services/accept_offer.rb [hunk 3]
```diff
@@ -46,6 +46,6 @@ def application_choices_awaiting_provider_decision
   end
 
   def unconditional_offer?
-    @application_choice.offer&.fetch('conditions', []).blank?
+    @application_choice.offer.conditions.none?
```
Comment (benswannack) [inline app/services/accept_offer.rb:49]: Can we just use `@application_choice.unconditional_offer?` here?

File: app/models/application_choice.rb (hunk 2 above)
Comment (despo) [reply app/models/application_choice.rb:81]: It was related to test data creation. Adding the offer association to the `with_offer` scope of the application choice factory was causing it to generate multiple application choices which as a result caused irrelevant specs who relied on a specific number of applciation choices being generated to fail, so I rolled it all back and decided to just leave it out for now.

File: app/services/accept_offer.rb (hunk 3 above)
Comment (despo) [reply app/services/accept_offer.rb:49]: No, because `unconditional_ofer?` returns false unless the application choice is in the recruited state. 

Something to tackle as part of the refactoring task.

File: app/views/candidate_interface/decisions/offer.html.erb [hunk 4]
```diff
@@ -30,10 +30,10 @@
       <p class="govuk-body">If you need help, you can speak to our <%= govuk_link_to 'teacher training advisers', t('get_into_teaching.url_get_an_advisor') %>.</p>
 
       <%= f.govuk_radio_buttons_fieldset :response, legend: { text: t('decisions.response.legend'), size: 'm' } do %>
-        <% if @application_choice.offer&.dig('conditions')&.present? %>
-          <%= f.govuk_radio_button :response, 'accept', label: { text: t('decisions.response.accept.label') }, link_errors: true %>
-        <% else %>
+        <% if @application_choice.unconditional_offer? %>
```
Comment (benswannack) [inline app/views/candidate_interface/decisions/offer.html.erb:33]: I don't think we can use this method here because of the `return false unless recruited?` line.

File: app/services/accept_offer.rb (hunk 3 above)
Comment (benswannack) [reply app/services/accept_offer.rb:49]: Ok, it's very confusing as is so we should address at some point

File: spec/forms/support_interface/conditions_form_spec.rb [hunk 5]
```diff
@@ -216,20 +216,26 @@
     end
 
     it 'reads standard and further conditions' do
-      application_choice = build(
-        :application_choice,
-        offer: { 'conditions' => ['Fitness to train to teach check', 'Get a haircut'] },
```
Comment (benswannack) [inline spec/forms/support_interface/conditions_form_spec.rb:221]: There are a few other instances of `'conditions' => '...'` in this spec. It would be nice to ensure we are testing the new behaviour everywhere and not relying on the old offer implementation

File: app/controllers/vendor_api/decisions_controller.rb (hunk 1 above)
Comment (benswannack) [reply app/controllers/vendor_api/decisions_controller.rb:70]: Maybe we could add some TODOs that mention the Trello card in which we'll add the association?

File: app/models/application_choice.rb (hunk 2 above)
Comment (benswannack) [reply app/models/application_choice.rb:81]: Will we add this back in the refactor card then? It's not too important since this does the same thing anyway

File: spec/forms/support_interface/conditions_form_spec.rb (hunk 5 above)
Comment (despo) [reply spec/forms/support_interface/conditions_form_spec.rb:221]: Removed

File: app/models/application_choice.rb (hunk 2 above)
Comment (despo) [reply app/models/application_choice.rb:81]: I don't know if it should be done as part of the refactor card. It may be simpler to wait until the offer field has been removed first. Maybe a small tech debt card.

File: app/controllers/vendor_api/decisions_controller.rb (hunk 1 above)
Comment (despo) [reply app/controllers/vendor_api/decisions_controller.rb:70]: I would reference from trello rather than from the code, as I don't think we do go through TODOs, so it's just going to end up being more noise that we ignore.

Comment (despo) [reply app/controllers/vendor_api/decisions_controller.rb:70]: https://trello.com/c/Cn4e6z2P/1588-add-offer-association-to-applicationchoice-model

**Review by benswannack — APPROVED** [review]
This looks good to me :) Probably worth waiting for another reviewer though as this is a large change!

File: app/components/candidate_interface/application_dashboard_course_choices_component.rb [hunk 6]
```diff
@@ -109,13 +109,13 @@ def interview_row(application_choice)
...
-        (application_choice.offer.present? && !application_choice.offer['conditions'].count.zero?)
+        (application_choice.offer.present? && application_choice.offer.conditions.any?)
 
       {
-        key: 'Condition'.pluralize(application_choice.offer['conditions'].count),
+        key: 'Condition'.pluralize(application_choice.offer.conditions.count),
         value: render(
           OfferConditionsReviewComponent.new(
-            conditions: application_choice.offer['conditions'],
+            conditions: application_choice.offer.conditions.map(&:text),
```
Comment (cwrw) [inline app/components/candidate_interface/application_dashboard_course_choices_component.rb:118]: might be nice to have a method in `offer` that does `conditions.map(&:text)` and we can invoke everywhere?

File: app/components/provider_interface/conditions_component.html.erb [hunk 7]
```diff
@@ -1,20 +1,26 @@
...
+          <tr class="govuk-table__row conditions-row">
+            <td class="govuk-table__cell"><%= condition.text %></td>
+            <td class="govuk-table__cell govuk-table__cell--numeric">
+              <% if condition.met? %>
+                <%= govuk_tag(text: 'Met', colour: 'green') %>
+              <% elsif condition.unmet? %>
+                <%= govuk_tag(text: 'Not met', colour: 'red') %>
+              <% else %>
+                <%= govuk_tag(text: 'Pending', colour: 'grey') %>
+              <% end %>
```
Comment (benswannack) [inline app/components/provider_interface/conditions_component.html.erb:20]: Just testing this locally in the deferred offer flow. This component is used when previewing a reinstated deferred offer. So this if condition needs to show the new condition status rather than the current one

**Review by cwrw — COMMENTED** [review]
looks great, amazing work @despo!

The only thing I found whilst testing:
- The unconditional offer flow isn't working as expected, I'm going through pending conditions instead of the application choice being instantly recruited

File: app/components/provider_interface/conditions_component.html.erb (hunk 7 above)
Comment (cwrw) [reply app/components/provider_interface/conditions_component.html.erb:20]: Yes got the same 👍

File: app/forms/provider_interface/reconfirm_deferred_offer_wizard.rb [hunk 8]
```diff
@@ -68,15 +68,19 @@ def course_option
     def modified_application_choice
       return unless application_choice
 
-      clone = application_choice.dup
-      clone.id = application_choice.id
```
Comment (despo) [inline app/forms/provider_interface/reconfirm_deferred_offer_wizard.rb:72]: Couldn't see why we had to use dup instead of clone, as we still assign the id to the application choice. It was also getting a lit bit more complex as we would then have to unecessarily do the same to offer conditions and then set the id, so switched to using `clone` instead.

Comment (benswannack) [reply app/forms/provider_interface/reconfirm_deferred_offer_wizard.rb:72]: I suppose `clone` might make a difference if we start using the original application form data to render the previous pages in the wizard flow. Updating the status of the `clone` also updates the status of the original.

Might need to revisit this when fixing that bug around cancelling/pressing back

File: app/forms/provider_interface/reconfirm_deferred_offer_wizard.rb [hunk 9]
```diff
@@ -68,15 +68,19 @@ def course_option
...
-                       clone.status_before_deferral
-                     end
-      clone
+      modified_application_choice = application_choice.clone
+      modified_application_choice.status = if conditions_status.present?
+                                             if conditions_met?
+                                               modified_application_choice.offer.conditions.each { |condition| condition.status = 'met' }
+                                               'recruited'
+                                             else
+                                               modified_application_choice.offer.conditions.each { |condition| condition.status = 'pending' }
```
Comment (benswannack) [inline app/forms/provider_interface/reconfirm_deferred_offer_wizard.rb:77]: Is it worth adding a small check to one of the existing system specs that covers this behaviour?

Comment (despo) [reply app/forms/provider_interface/reconfirm_deferred_offer_wizard.rb:77]: updated existing spec to verify condition status

**Review by cwrw — APPROVED** [review]
Looks good, previous problem with deferrals fixed, and unconditional wasn't an issue to begin with. Thanks for this @despo

**Review by benswannack — APPROVED** [review]
Nice!
